extensions/sql-database-projects/src/controllers/projectController.ts

```diff
@@ -32,11 +45,158 @@ export class ProjectsController {
+		if (!treeNode) {
+			// TODO: prompt for which (currently-open) project when invoked via command pallet
+			throw new Error('TODO: prompt for which project when invoked via command pallet');
+		}
+
+		if (treeNode.root instanceof ProjectRootTreeItem) {
+			return (treeNode.root as ProjectRootTreeItem).project;
+		}
+		else {
+			throw new Error('"Add item" command invoked from unexpected location: ' + treeNode.uri.path);
+		}
+	}
+
+	private async promptForNewObjectName(itemType: templateMap.ProjectScriptType, _project: Project): Promise<string | undefined> {
+		// TODO: ask project for suggested name that doesn't conflict
+		const suggestedName = itemType.friendlyName.replace(new RegExp('\s', 'g'), '') + '1';
+
+		const itemObjectName = await vscode.window.showInputBox({
+			prompt: constants.newObjectNamePrompt(itemType.friendlyName),
+			value: suggestedName,
+		});
+
+		return itemObjectName;
+	}
+
+	private prependContextPath(treeNode: BaseProjectTreeItem, objectName: string): string {
+		if (treeNode instanceof FolderNode) {
+			return path.join(utils.trimUri(treeNode.root.uri, treeNode.uri), objectName);
+		}
+		else {
+			return objectName;
+		}
+	}
+
+	public async addFolderPrompt(treeNode: BaseProjectTreeItem) {
+		const project = this.getProjectContextFromTreeNode(treeNode);
+		const newFolderName = await this.promptForNewObjectName(new templateMap.ProjectScriptType(templateMap.folder, 'Folder', ''), project);
+
+		if (!newFolderName) {
+			return; // user cancelled
+		}
+
+		const relativeFolderPath = this.prependContextPath(treeNode, newFolderName);
+
+		await project.addFolderItem(relativeFolderPath);
+
+		this.refreshProjectsTree();
+	}
+
+	public async addItemPrompt(treeNode: BaseProjectTreeItem, itemTypeName?: string) {
+		const project = this.getProjectContextFromTreeNode(treeNode);
+
+		if (!itemTypeName) {
+			let itemFriendlyNames: string[] = [];
+
+			for (const itemType of templateMap.projectScriptTypes) {
+				itemFriendlyNames.push(itemType.friendlyName);
+			}
+
+			itemTypeName = await vscode.window.showQuickPick(itemFriendlyNames, {
+				canPickMany: false
+			});
+
+			if (!itemTypeName) {
+				return; // user cancelled
+			}
+		}
+
+		const itemType = templateMap.projectScriptTypeMap[itemTypeName.toLocaleLowerCase()];
+		const itemObjectName = await this.promptForNewObjectName(itemType, project);
+
+		if (!itemObjectName) {
+			return; // user cancelled
+		}
+
+		// TODO: file already exists?
+
+		const newFileText = this.macroExpansion(itemType.templateScript, { 'OBJECT_NAME': itemObjectName });
+		const relativeFilePath = this.prependContextPath(treeNode, itemObjectName + '.sql');
+
+		const newEntry = await project.addScriptItem(relativeFilePath, newFileText);
+
+		vscode.commands.executeCommand('vscode.open', newEntry.fsUri);
 
 		this.refreshProjectsTree();
 	}
 
+	private macroExpansion(template: string, macroDict: Record<string, string>): string {
```

Comment:
can you please add some tests for this file?

Reply:
Added